File: Problems/Arrays/Best-Time-to-Buy-and-Sell-Stock/Best-Time-to-Buy-and-Sell-Stock.cpp
#include <iostream>
#include <vector>
#include <climits>

// Using dynamic programming technique - Time: O(n)

class Solution {
public:
    int maxProfit(std::vector<int> &prices)
    {
        int buy = INT_MIN, sell = 0;
        
        for (int price : prices)
        {
            buy = std::max(buy, -price);
            sell = std::max(sell, buy + price);
        }

        return sell;
    }
};

int main()
{
    std::vector<int> prices = {7, 1, 5, 3, 6, 4};

    // Print input
    std::cout << "Input: prices = [";
    for (const auto &price : prices)
    {
        std::cout << price << "";
        if (&price != &prices.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    Solution sol;
    int maxProfit = sol.maxProfit(prices);

    // Print output
    std::cout << "Output: " << maxProfit << std::endl;

    return 0;
}

File: Problems/Arrays/Best-Time-to-Buy-and-Sell-Stock/Best-Time-to-Buy-and-Sell-Stock.cs
using System;

// Using dynamic programming technique - Time: O(n)

public class Solution
{
    public int MaxProfit(int[] prices)
    {
        int buy = int.MinValue, sell = 0;

        foreach (int price in prices)
        {
            buy = Math.Max(buy, -price);
            sell = Math.Max(sell, buy + price);
        }

        return sell;
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[] prices = { 7, 1, 5, 3, 6, 4 };

        // Print input
        Console.Write("Input: prices = [");
        foreach (int price in prices)
        {
            Console.Write(price + "");
            if (price != prices[prices.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("]");

        Solution solution = new Solution();
        int maxProfit = solution.MaxProfit(prices);

        // Print output
        Console.WriteLine("Output: " + maxProfit);
    }
}